bftengine/tests/simpleStorage/FileStorage.cpp

```diff
@@ -0,0 +1,250 @@
+  LOG_INFO(logger_,
+           "FileStorage::initMaxSizeOfObjects Maximum size of objects is "
+               << maxObjectsSize << ", file size is " << maxFileSize);
+  ftruncate(fileno(dataStream_), maxFileSize);
+  writeFileMetadata();
+  LOG_INFO(logger_, "" << *objectsMetadata_);
+}
+
+void FileStorage::verifyFileMetadataSetup() const {
+  if (!objectsMetadata_ || !(objectsMetadata_->getObjectsNum())) {
+    LOG_FATAL(logger_, "FileStorage::verifyFileMetadataSetup "
+        << METADATA_IS_NOT_SET_PROPERLY);
+    throw runtime_error(METADATA_IS_NOT_SET_PROPERLY);
+  }
+}
+
+void FileStorage::verifyOperation(uint16_t objectId, uint32_t dataLen,
+                                  char *buffer) const {
+  verifyFileMetadataSetup();
+  MetadataObjectInfo *objectInfo = objectsMetadata_->getObjectInfo(objectId);
+  if (!objectInfo || objectId > objectsMetadata_->getObjectsNum() - 1 ||
+      !dataLen || dataLen > objectInfo->maxSize || !buffer) {
+    LOG_FATAL(logger_, "FileStorage::verifyOperation " << WRONG_PARAMETER);
+    throw runtime_error(WRONG_PARAMETER);
+  }
+}
+
+void FileStorage::handleObjectWrite(uint16_t objectId, void *dataPtr,
+                                    uint32_t objectSize) {
+  MetadataObjectInfo *objectInfo = objectsMetadata_->getObjectInfo(objectId);
+  if (objectInfo) {
+    write(dataPtr, objectInfo->offset, objectSize, 1,
+          FAILED_TO_WRITE_OBJECT);
+    LOG_INFO(logger_, "FileStorage::handleObjectWrite "
+        << objectInfo->toString());
+    // Update the file metadata with a real object size
+    objectInfo->realSize = objectSize;
+    objectsMetadata_->setObjectInfo(*objectInfo);
+    updateFileObjectMetadata(*objectInfo);
+  }
+}
+
+void FileStorage::handleObjectRead(uint16_t objectId, char *outBufferForObject,
+                                   uint32_t &outActualObjectSize) {
+  MetadataObjectInfo *objectInfo = objectsMetadata_->getObjectInfo(objectId);
+  if (objectInfo) {
+    read(outBufferForObject, objectInfo->offset, objectInfo->realSize, 1,
+         FAILED_TO_READ_OBJECT);
+    outActualObjectSize = objectInfo->realSize;
+    LOG_INFO(logger_, "FileStorage::handleObjectRead "
+        << objectInfo->toString());
+  }
+}
+
+void FileStorage::read(uint16_t objectId, uint32_t bufferSize,
+                       char *outBufferForObject,
+                       uint32_t &outActualObjectSize) {
+  LOG_INFO(logger_, "FileStorage::read objectId="
+      << objectId << ", bufferSize=" << bufferSize);
+  verifyOperation(objectId, bufferSize, outBufferForObject);
+  lock_guard<mutex> lock(ioMutex_);
+  handleObjectRead(objectId, outBufferForObject, outActualObjectSize);
+}
+
+void FileStorage::atomicWrite(uint16_t objectId, char *data,
+                              uint32_t dataLength) {
+  LOG_INFO(logger_, "FileStorage::atomicWrite objectId="
+      << objectId << ", dataLength=" << dataLength);
+  verifyOperation(objectId, dataLength, data);
+  lock_guard<mutex> lock(ioMutex_);
```

Comment:
Move the lock guard above `verifyOperation` please.

Reply:
Done.